### PR TITLE
Bulk Downloads Invoices 

### PR DESCRIPTION
**Pull request by vipulnsward**
Moved from https://github.com/saeloun/miru-web/pull/769

## Notion card
https://www.notion.so/saeloun/Backend-Bulk-Invoice-download-da485b6db7d64df1b4fefeb88740e3e0

## Summary
- Added API for bulk invoice download. 
- This API accepts a download id and the invoice ids. 
- It starts a background job for creating zip of asked invoice ids and also creates a channel with given download id.
- Once the invoices zip is ready, it publishes that zip on the download id channel.

### CODE REVIEW

**Comment by vipulnsward** [comment]
@shalapatil bundle fails locally, we need to check, I can sync up on it Monday if needed

**Comment by xrimpy** [comment]
Bundle is failing because there is a mismatch between Gemfile and Gemfile.lock

Following lines have been omitted  from Gemfile.lock even though the rspec-sidekiq exists in Gemfile but not in the Lockfile. 
```
    rspec-sidekiq (3.1.0)
      rspec-core (~> 3.0, >= 3.0.0)
      sidekiq (>= 2.4.0)
```

File: app/models/invoice.rb [hunk 1]
```diff
@@ -103,6 +103,24 @@ def unit_amount(base_currency)
...
+
+  def temp_pdf(company_logo, root_url)
+    file = Tempfile.new()
+    InvoicePayment::PdfGeneration.process(
+      self,
+      company_logo,
+      root_url,
+      file.path)
+    file
+  end
```
Comment (shalapatil) [inline app/models/invoice.rb:122]: temp_pdf implementation is changed

File: app/services/bulk_invoice_download_service.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  def process
+    zip_invoices
+  end
+
+  private
+
+    def zip_invoices
+      zipper = Zipper.new(invoices_temp_pdf_data)
+      zipper.zip
+      file_uploader = FileUploader.new(zipper.file, "Invoices_" + Time.now.to_i.to_s + ".zip")
```
Comment (shalapatil) [inline app/services/bulk_invoice_download_service.rb:21]: file_uploader is new

File: app/services/invoice_payment/pdf_generation.rb [hunk 3]
```diff
@@ -2,11 +2,12 @@
 
 module InvoicePayment
   class PdfGeneration < ApplicationService
-    def initialize(invoice, company_logo, root_url)
+    def initialize(invoice, company_logo, root_url, filepath = nil)
       @invoice = invoice
```
Comment (shalapatil) [inline app/services/invoice_payment/pdf_generation.rb:6]: this is changed

Comment (shalapatil) [reply app/services/invoice_payment/pdf_generation.rb:6]: all the changes from this file

File: lib/file_uploader.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
```
Comment (shalapatil) [inline lib/file_uploader.rb:1]: this is newly added file

File: spec/services/bulk_invoice_download_service_spec.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  let!(:invoice2) { create(:invoice, client:) }
+  let(:invoice3) { create(:invoice, client:) }
+
+  describe "#process" do
+    before do
+      # While running specs, we don't want url and to access url, we need to set ActiveStorage url
+      allow_any_instance_of(ActiveStorage::Blob).to receive(:url).and_return("")
+
+      described_class.new([invoice1.id, invoice2.id], nil, "").process
+
```
Comment (shalapatil) [inline spec/services/bulk_invoice_download_service_spec.rb:20]: this spec implementation is changed as service response has changed

**Pull request merged**